game-core/src/main/java/games/strategy/triplea/ui/PlacePanel.java

```diff
@@ -94,13 +107,106 @@ public void territorySelected(final Territory territory, final MouseDetails e) {
+      unitsToPlacePanel.setVisible(false);
+      data.addGameDataEventListener(GameDataEvent.GAME_STEP_CHANGED, this::updateStep);
+    } else {
+      unitsToPlacePanel = new SimpleUnitPanel(map.getUiContext());
+    }
     leftToPlaceLabel.setText("Units left to place:");
   }
 
+  private void updateStep() {
+    final Collection<UnitCategory> unitsToPlace;
+    final GameData data = getData();
+    data.acquireReadLock();
+    try {
+      final GameStep step = data.getSequence().getStep();
+      if (step == null) {
+        return;
+      }
+      // Note: This doesn't use getCurrentPlayer() as that may not be updated yet.
+      final PlayerId player = step.getPlayerId();
+      unitsToPlace = getUpdatedUnitsToPlace(lastPlayer, player, step);
+      lastPlayer = player;
+    } finally {
+      data.releaseReadLock();
+    }
+
+    SwingUtilities.invokeLater(
+        () -> {
+          JSplitPane splitPane = null;
+          final Component parent = unitsToPlacePanel.getParent();
+          if (parent instanceof JSplitPane) {
+            splitPane = ((JSplitPane) parent);
+          }
+          if (unitsToPlace != null) {
+            // Show the panel.
+            unitsToPlacePanel.setUnitsFromCategories(unitsToPlace);
+            if (!unitsToPlacePanel.isVisible()) {
+              unitsToPlacePanel.setVisible(true);
+              // If we're making the panel visible, also make the split
+              // pane divider visible and maximize.
+              if (splitPane != null) {
+                splitPane.setDividerSize(8);
+                if (!splitPaneWasMinimized) {
+                  splitPane.resetToPreferredSizes();
+                }
+              }
+            }
+          } else {
+            // Hide the panel.
+            if (splitPane != null) {
+              // If the panel was visible before, remember whether the split pane was
+              // minimized so that we don't open it up again the next time we show it.
+              if (unitsToPlacePanel.isVisible()) {
+                // Note: We don't use getMaximumDividerLocation() because that takes into account
+                // the preferred size of the unit panel and therefore when not minimized, it may
+                // still be at its "maximum location".
+                final var dividerBottom =
+                    splitPane.getDividerLocation() + splitPane.getDividerSize();
+                splitPaneWasMinimized = (dividerBottom == splitPane.getHeight());
+              }
+              // Set the divider size to 0 to hide the divider UI.
+              splitPane.setDividerSize(0);
+            }
+            unitsToPlacePanel.setVisible(false);
+            unitsToPlacePanel.removeAll();
+          }
+        });
+  }
+
+  private @Nullable Collection<UnitCategory> getUpdatedUnitsToPlace(
+      final PlayerId lastPlayer, final PlayerId player, final GameStep step) {
+    Collection<UnitCategory> unitCategories = null;
```

Comment:
No, because null has the meaning of "don't show the panel" whereas an empty list means "show an empty panel". The two are distinct because it's important to note when there is simply no purchase (or no more units to place), as opposed to we shouldn't show the panel because it's before purchase phase.

Reply:
Okay, but we should tread carefully. Any variable where the `null` value means one thing, and the value of the variable when set means another, is a SRP violation. `Optional<Collection<Unitcategory>>` avoids at least the `null` and makes this double meaning explicit.

Reply:
I'm now using a "boolean showUnitsToPlace;". Let me know if you still prefer an Optional instead of the boolean.